url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/bufferManager/BufferManager.java

```diff
@@ -0,0 +1,160 @@
+package ru.spbau.tinydb.bufferManager;
+
+import ru.spbau.tinydb.common.DBException;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Sergey on 27.10.2014.
+ */
+public class BufferManager {
+	// TODO add some header(magic constants, version, other very useful info)
+	// TODO add check sum(methods write/load page)
+	
+    public final static long PAGE_SIZE = 1 << 12; // 4KB
+    private final static int BUFFER_SIZE = 1 << 10; // in PAGE_SIZE's
+    public final static int METAINF_FIRST_PAGE = 1;
+    public final static int INDEXES_FIRST_PAGE = 2;
+
+    private final RandomAccessFile dbFile;
+    private final ByteBuffer buffer = ByteBuffer.allocate((int) (PAGE_SIZE * BUFFER_SIZE));
+    // 'i' page is used now if 'i' bit is set
+    private final BitSet bufferIsUsed = new BitSet(BUFFER_SIZE);
+    // 'i' page was recently used if 'i' bit is set
+    private final BitSet bufferWasUsed = new BitSet(BUFFER_SIZE);
+    // 'i' page should be written to disk if 'i' bit is set
+    private final BitSet bufferChanged = new BitSet(BUFFER_SIZE);
+    private final int[] usedPages = new int[BUFFER_SIZE];
+    private final Map<Integer, Integer> pageIdxToBufferPos = new HashMap<>(); // pageIndex -> buffer position
+    private final int[] bufferPosToPageIdx = new int[BUFFER_SIZE]; // buffer position -> pageIndex  
+    private int currentPos = -1;
+    private int length;
+
+    public BufferManager(String dbFileName) throws DBException {
+        this(new File(dbFileName));
+    }
+
+    public BufferManager(File dbFile) throws DBException {
+        try {
+            this.dbFile = new RandomAccessFile(dbFile, "rw");
+        } catch (FileNotFoundException e) {
+            throw new DBException(e);
+        }
+
+        init();
+    }
+
+    private void init() throws DBException {
+        Arrays.fill(bufferPosToPageIdx, -1);
+        try {
+            length = (int) (dbFile.length() / PAGE_SIZE);
+            if(length == 0) {
+                increaseLength();
+                //writeHeader(there is no header now)
+                increaseLength(); // for MetaInf Page
+                increaseLength(); // for Indexes Page
+            } else {
+                //loadHeader(there is no header now)
+            }
+        } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)